File: android/app/src/test/kotlin/jp/golia/mailrs/wire/DateChipTest.kt
package jp.golia.mailrs.wire

import java.util.Locale
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class DateChipTest {
    private fun s(date: String, datetime: String? = null, text: String = "written") =
        Wire.DateSuggestion(date = date, datetime = datetime, text = text)

    /**
     * The row from the 2026-08-21 report: three days written three
     * ways. What must not differ afterwards is the pattern.
     */
    @Test
    fun `three formats become one shape`() {
        val labels =
            listOf(
                    s("2026-08-21", text = "Aug 21 2026"),
                    s("2026-08-20", text = "2026-08-20"),
                    s("2026-08-19", text = "2026-08-19"),
                )
                .map { DateChip.label(it, Locale.UK) }
        val shapes = labels.map { it.replace(Regex("\\d+"), "#") }.toSet()
        assertEquals(1, shapes.size)
        assertEquals(3, labels.toSet().size)
    }

    /** A bare date read as an instant would render as the day before. */
    @Test
    fun `the day is the local day`() {
        assertTrue(DateChip.label(s("2026-08-21"), Locale.UK).contains("21"))
    }

    @Test
    fun `an hour shows only when one was written`() {
        val withTime = DateChip.label(s("2026-08-25", "2026-08-25T14:00:00"), Locale.UK)
        val allDay = DateChip.label(s("2026-08-25"), Locale.UK)
        assertNotEquals(withTime, allDay)
        assertTrue(withTime.length > allDay.length)
    }

    /** Unparseable input falls back rather than inventing a day. */
    @Test
    fun `nonsense falls back`() {
        // No date at all: the writer's words are all there is.
        assertEquals("written", DateChip.label(s("not-a-date"), Locale.UK))
        // A date but an unreadable clock: the day, and no invented hour.
        assertEquals(
            DateChip.label(s("2026-08-25"), Locale.UK),
            DateChip.label(s("2026-08-25", "garbage"), Locale.UK),
        )
    }
}

File: web/src/lib/__tests__/date-chip.test.ts
import { describe, expect, it } from 'vitest'

import { chipLabel } from '../date-chip'

describe('chipLabel', () => {
  it('gives one shape to dates the writer wrote differently', () => {
    // The row from the 2026-08-21 report: three formats, three days.
    const written = [
      { date: '2026-08-21', datetime: null, text: 'Aug 21 2026' },
      { date: '2026-08-20', datetime: null, text: '2026-08-20' },
      { date: '2026-08-19', datetime: null, text: '2026-08-19' },
    ]
    const labels = written.map(chipLabel)
    // Same shape for all three — the writer's form no longer leaks into
    // the row. The weekday and day differ, of course; what must not
    // differ is the pattern, so that is what is asserted.

    for (const l of labels)
      expect(l.replace(/[A-Za-z\u4e00-\u9fff]+/g, 'X')).toMatch(/^[X, ]*\d{1,2}[X, ]*$/)
    expect(new Set(labels).size).toBe(3)
  })

  it('reads the day in the local calendar, not UTC', () => {
    // `new Date('2026-08-21')` is UTC midnight and renders as the 20th
    // anywhere west of Greenwich. Built from parts, it does not.
    expect(chipLabel({ date: '2026-08-21', datetime: null })).toContain('21')
  })

  it('carries the hour when one was written, and not when it was not', () => {
    const withTime = chipLabel({ date: '2026-08-25', datetime: '2026-08-25T14:00:00' })
    const allDay = chipLabel({ date: '2026-08-25', datetime: null })
    expect(withTime.length).toBeGreaterThan(allDay.length)
    expect(withTime.startsWith(allDay)).toBe(true)
  })

  it('falls back to the day when the time is unreadable', () => {
    expect(chipLabel({ date: '2026-08-25', datetime: '2026-08-25' })).toBe(
      chipLabel({ date: '2026-08-25', datetime: null })
    )
  })
})
